contracts/stacks/contracts/xcall/xcall-impl.clar

```diff
@@ -0,0 +1,708 @@
+    (var-set sn-counter (+ current-sn u1))
+    (ok (+ current-sn u1))
+  )
+)
+
+(define-private (get-next-req-id)
+  (let (
+    (current-id (var-get req-id-counter))
+  )
+    (var-set req-id-counter (+ current-id u1))
+    (ok (+ current-id u1))
+  )
+)
+
+(define-private (validate-network-address (address (string-ascii 257)))
+  (match (index-of? address "/")
+    index 
+      (let 
+        (
+          (network-id-in (slice? address u0 index))
+          (account (slice? address (+ index u1) (len address)))
+        )
+        (and 
+          (is-some network-id-in)
+          (is-some account)
+          (> (len (unwrap! network-id-in false)) u0)
+          (> (len (unwrap! account false)) u0)
+        )
+      )
+    false
+  )
+)
+
+(define-private (parse-network-address (address (string-ascii 257)))
+  (if (validate-network-address address)
+    (match (index-of? address "/")
+      index 
+        (let 
+          (
+            (network-id-in (unwrap-panic (as-max-len? (unwrap-panic (slice? address u0 index)) u128)))
+            (account (unwrap-panic (slice? address (+ index u1) (len address))))
+          )
+          (ok {network-id: network-id-in, account: account})
+        )
+      ERR_INVALID_NETWORK_ADDRESS
+    )
+    ERR_INVALID_NETWORK_ADDRESS
+  )
+)
+
+(define-public (set-trusted-protocols (nid (string-ascii 128)) (protocols (list 10 (string-ascii 128))))
+  (begin
+    (asserts! (is-admin) ERR_NOT_ADMIN)
+    (ok (map-set trusted-protocols { nid: nid } { protocols: protocols }))
+  )
+)
+
+(define-private (emit-call-message-received-event (from (string-ascii 128)) (to (string-ascii 128)) (sn uint) (req-id uint) (data (buff 2048)))
+  (print
+    {
+      event: "CallMessage",
+      from: from,
+      to: to,
+      sn: sn,
+      req-id: req-id,
+      data: data,
+    }
+  )
+)
+
+(define-private (emit-call-executed-event (req-id uint) (code uint) (message (string-ascii 128)))
+  (print
+    {
+      event: "CallExecuted",
+      req-id: req-id,
+      code: code,
+      msg: message
+    }
+  )
+)
+
+(define-private (emit-call-message-sent-event (from principal) (to (string-ascii 128)) (sn uint))
+  (print
+    {
+      event: "CallMessageSent",
+        from: tx-sender,
+        to: to,
+        sn: sn,
+    }
+  )
+)
+
+(define-private (emit-response-message-event (sn uint) (code uint))
+  (print 
+    {
+      event: "ResponseMessage",
+      sn: sn,
+      code: code
+    }
+  )
+)
+
+(define-private (emit-rollback-message-received-event (sn uint))
+  (print 
+    {
+      event: "RollbackMessage",
+      sn: sn,
+    }
+  )
+)
+
+(define-private (emit-rollback-executed-event (sn uint))
+  (print 
+    {
+      event: "RollbackExecuted",
+      sn: sn
+    }
+  )
+)
+
+(define-read-only (get-default-connection (nid (string-ascii 128)))
+  (match (map-get? default-connections { nid: nid })
+    connection (ok (some connection))
+    ERR_NO_DEFAULT_CONNECTION)
+)
+
+(define-public (send-call 
```

Comment:
I think message types have not been done so far right?

Reply:
https://github.com/icon-project/xcall-multi/tree/main/contracts/javascore/xcall-lib/src/main/java/foundation/icon/xcall/messages